tests/testthat/helper-expect-metadata.R

```diff
@@ -0,0 +1,258 @@
+  expect_equal(filters[[5]], "sex", ignore_attr = TRUE)
+  expect_equal(filters[[6]], "age_group", ignore_attr = TRUE)
+  expect_equal(filters[[7]], "source", ignore_attr = TRUE)
+}
+
+mock_filter_types <- function(ids) {
+  lapply(ids, function(id) {
+    list(
+      id = scalar(id),
+      column_id = scalar(paste0("col_", id)),
+      options = list(
+        list(
+          id = scalar("opt1"),
+          label = scalar("Option 1")
+        ),
+        list(
+          id = scalar("opt2"),
+          label = scalar("Option 2")
+        )
+      )
+    )
+  })
+}
+
+## Test validity of the metadata, namely that ids used in any effect which
+## reference filters actually exist in the filter types
+expect_valid_metadata <- function(metadata) {
+  filter_ids <- lapply(metadata$filterTypes, "[[", "id")
+  filter_types <- setNames(metadata$filterTypes, filter_ids)
+
+  plot_names <- names(metadata$plotSettingsControl)
+  for (plot in plot_names) {
+    default_effects <- metadata$plotSettingsControl[[plot]]$defaultEffect
+    if (length(metadata$plotSettingsControl[[plot]]$plotSettings) == 0) {
+      validate_effects(filter_types, default_effects, list(),
+                       paste0("plot: '", plot, "', default effects"))
+    } else {
+      for (plot_setting in metadata$plotSettingsControl[[plot]]$plotSettings) {
+        for (option in plot_setting$options) {
+          validate_effects(
+            filter_types,
+            option$effect,
+            default_effects,
+            paste0("plot: '",  plot,
+                  "', setting: '", plot_setting$id,
+                  "', option: '", option$id, "'"))
+        }
+      }
+    }
+  }
+  invisible(TRUE)
+}
+
+#' Validate effects
+#'
+#' Effects have filterId and stateFilterId
+#' stateFilterId is the unique ID for the dropdown
+#' filterId is the unique ID of the FilterType, used to identify options
+#' this is because there can be multiple dropdowns using the same underlying
+#' data e.g. colour and size indicator in the bubble plot
+#' setFilter effects have the mapping between filterId and stateFilterId
+#' all other effects refer only to the stateFilterId
+#'
+#' @param filter_types Named filter types
+#' @param effects Effects to check
+#' @param default_effects Default effects applied for every plot option
+#'   must be validated together as effects are additive
+#' @param context Additional context appended to any test failure messages
+#'
+#' @return Nothing, called for side effect
+#' @noRd
+validate_effects <- function(filter_types, effects, default_effects, context) {
+  ## Effects are additive, except for setFilters where effects on an
+  ## individual filter override the defaults
+  set_filters <- effects$setFilters
+  if (length(effects$setFilters) == 0) {
+    set_filters <- default_effects$setFilters
+  }
+  expect(length(set_filters) > 0,
+         paste(context, "setFilters must be set in plot setting or",
+               "default settings."))
```

Comment:
while this is currently true for all plots, this is not necessarily true, feel free to leave this in, but you can technically do more with the metadata.

For example, say there was a plot control `A` and a plot control `B` and a default effect `C`. You can have it so that only `A` contains a `setFilters` effect for all its settings and `B`, `C` do not contain a `setFilters`. Then when your test gets to plot control `B`, it will fail this test but every option from plot control `A` still specifies `setFilters` so everything in the UI would be fine

if you want to properly test this behaviour i would go one level up, pull our combined `setFilters` (combined from all combinations of all available plot controls + default effect) and test that none of those combined `setFilters` are empty. that would test it out properly but up to you whether you want to do this or not

Reply:
Thanks, added a note about this (feeling a bit lazy!)

Reply:
no problem!